### PR TITLE
[Feat] Autoaugment functions.

### PR DESCRIPTION
**Pull request by shijianjian**
### Description
A few sentences describing the changes proposed in this pull request.

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/color/adjust.py [hunk 1]
```diff
@@ -225,6 +227,169 @@ def adjust_brightness(input: torch.Tensor,
...
+        if factor == 1.:
+            return input2
+        diff = (input2 - input1) * factor
+        res = input1 + diff
+        if factor > 0. and factor < 1.:
+            return res
+        return torch.clamp(res, 0, 1)
+    if isinstance(factor, (float)) or len(factor) == 0:
+        return _blend_one(input, result, factor)
+    return torch.stack([_blend_one(input[i], result[i], factor[i]) for i in range(len(factor))])
```
Comment (edgarriba) [inline kornia/color/adjust.py:390]: For sharpness I found this much clear to me
https://stackoverflow.com/questions/4993082/how-to-sharpen-an-image-in-opencv

We support gaussian blur and add weighted

Maybe we could support both implementations by adding a flag `method` to the function signature. 

```python
if method == 'pillow':
  return _sharpen_pil(...)
elif method == 'default':
   return _sharpen(...)
else:
   raise ERROR
```

This could be applied to the other methods with different possible implementations. However we should find what are the actual methods names not names of the libraries implementing. /cc @ducha_aiki @krrish94

Comment (shijianjian) [reply kornia/color/adjust.py:417]: I think it is better to stick with one method. Or we shall support multi-backends globally but in a decent manner somehow, not rely on if else in each method.

**Comment by shijianjian** [comment]
One thing to mention here is, methods like posterize rely on uint8 to perform a bit-wise shift. I think it might be some issues towards differentiability. We might need to develop a differentiable version of kornia.LUT though I am not quite sure about it.

File: kornia/color/adjust.py (hunk 1 above)
Comment (edgarriba) [reply kornia/color/adjust.py:417]: it's quite common in computer vision libraries to have different methods solving same problem/task.

Comment (shijianjian) [reply kornia/color/adjust.py:419]: I mean it is better to have a global switch like ```kornia.backend = 'PIL'``` or ```kornia.backend = 'OpenCV'``` for a global behaviour control as well.

File: kornia/color/adjust.py [hunk 2]
```diff
@@ -225,6 +227,248 @@ def adjust_brightness(input: torch.Tensor,
...
+        raise TypeError(f"Input type is not a torch.Tensor. Got {type(input)}")
+
+    if not isinstance(thresholds, (float, torch.Tensor,)):
+        raise TypeError(f"The factor should be either a float or torch.Tensor. "
+                        f"Got {type(thresholds)}")
+
+    if isinstance(thresholds, float):
+        thresholds = torch.tensor(thresholds)
+
+    if additions is not None:
```
Comment (edgarriba) [inline kornia/color/adjust.py:283]: is None same as adding zeros ? actually default value is 0, right ?

File: kornia/color/adjust.py [hunk 3]
```diff
@@ -225,6 +227,248 @@ def adjust_brightness(input: torch.Tensor,
...
+
+    if not isinstance(thresholds, (float, torch.Tensor,)):
+        raise TypeError(f"The factor should be either a float or torch.Tensor. "
+                        f"Got {type(thresholds)}")
+
+    if isinstance(thresholds, torch.Tensor) and len(thresholds.shape) != 0:
+        assert input.size(0) == len(thresholds) and len(thresholds.shape) == 1, \
+            f"threshholds must be a 1-d vector of shape ({input.size(0)},). Got {thresholds}"
+        # TODO: I am not happy about this line, but no easy to do batch-wise operation
+        thresholds = torch.stack([x.expand(*input.shape[1:]) for x in thresholds])
```
Comment (edgarriba) [inline kornia/color/adjust.py:252]: if at this point we have an homogeneous shape BxCxHxW we can dot `thresholds = threshold.view(-1, 1, 1, 1)` then broadcasting will work. It's extremely ugly but this might work `thresholds.view(-1, *[1 for _ in range(len(data.shape)-1)])`

File: kornia/color/adjust.py [hunk 4]
```diff
@@ -225,6 +227,248 @@ def adjust_brightness(input: torch.Tensor,
...
+    _bits = bits.flatten()
+    for i in range(input.shape[0]):
+        res.append(_posterize_one(_input[i], _bits[i]))
+    return torch.stack(res, dim=0).reshape(*input.shape)
+
+
+def sharpness(input: torch.Tensor, factor: Union[float, torch.Tensor]) -> torch.Tensor:
+    r"""Implements Sharpness function from PIL using torch ops.
+    Args:
+        input (torch.Tensor): image or batched images for sharpness.
```
Comment (edgarriba) [inline kornia/color/adjust.py:373]: format and input shapes

File: kornia/color/adjust.py [hunk 5]
```diff
@@ -225,6 +227,248 @@ def adjust_brightness(input: torch.Tensor,
...
+        res = input1 + diff
+        if factor > 0. and factor < 1.:
+            return res
+        return torch.clamp(res, 0, 1)
+    if isinstance(factor, (float)) or len(factor.size()) == 0:
+        return _blend_one(input, result, factor)
+    return torch.stack([_blend_one(input[i], result[i], factor[i]) for i in range(len(factor))])
+
+
+def equalize(input):
```
Comment (edgarriba) [inline kornia/color/adjust.py:422]: typing

File: kornia/color/adjust.py [hunk 6]
```diff
@@ -225,6 +227,248 @@ def adjust_brightness(input: torch.Tensor,
...
+    """Implements Equalize function from PIL using PyTorch ops based on:
+    https://github.com/tensorflow/tpu/blob/master/models/official/efficientnet/autoaugment.py#L352
+    Args:
```
Comment (edgarriba) [inline kornia/color/adjust.py:425]: input/output sizes

File: kornia/utils/image.py [hunk 7]
```diff
@@ -24,24 +26,50 @@ def image_to_tensor(image: np.ndarray, keepdim: bool = True) -> torch.Tensor:
...
-    elif len(input_shape) == 3:
-        # (H, W, C) -> (C, H, W)
-        tensor = tensor.permute(2, 0, 1)
-    elif len(input_shape) == 4:
-        # (B, H, W, C) -> (B, C, H, W)
-        tensor = tensor.permute(0, 3, 1, 2)
-        keepdim = True  # no need to unsqueeze
-    else:
-        raise ValueError(
-            "Cannot process image with shape {}".format(input_shape))
```
Comment (edgarriba) [inline kornia/utils/image.py:42]: don't remove this piece of code. Seems pretty clear what is doing. What happens later it's not

File: kornia/utils/image.py [hunk 8]
```diff
@@ -24,24 +26,50 @@ def image_to_tensor(image: np.ndarray, keepdim: bool = True) -> torch.Tensor:
...
 
-    return tensor.unsqueeze(0) if not keepdim else tensor
+def to_bchw(tensor: torch.Tensor, color_channel_num: Optional[int] = None) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/utils/image.py:46]: I don't understand why this function. The name isn't appropriate and the docs + code readability don't help at all. I wouldn't add this to the main functionality. If it's a util for a specific transform make it first there, later if we see that can be useful or generalized we can discuss how adapt core functions to be more generic.

File: kornia/augmentation/augmentation.py [hunk 9]
```diff
@@ -21,6 +21,16 @@
 
 
 class AugmentationBase(nn.Module):
+    r"""AugmentationBase base class for customized augmentation implementations. For any augmentation,
+    the implementation of "generate_parameters" and "apply_transform" are required while the
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:25]: If so, I think we shall add a proper deprecation warning to do this breaking change. It is better to be implemented in a separate PR.

File: kornia/augmentation/augmentation.py [hunk 10]
```diff
@@ -753,10 +763,216 @@ def __repr__(self) -> str:
...
+        >>> solarize(input)
+        tensor([[[[0.4132, 0.1412, 0.1790, 0.2226, 0.3980],
+                  [0.2754, 0.4194, 0.0130, 0.4538, 0.2771],
+                  [0.4394, 0.4923, 0.1129, 0.2594, 0.3844],
+                  [0.3909, 0.2118, 0.1094, 0.2516, 0.3728],
+                  [0.2278, 0.0000, 0.4876, 0.0353, 0.5100]]]])
+    """
+
+    def __init__(
+        self, thresholds: FloatUnionType = 0.1, additions: FloatUnionType = 0.1,
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:806]: Reasonable. It will be included in the next PR.

File: kornia/augmentation/augmentation.py [hunk 11]
```diff
@@ -753,10 +763,216 @@ def __repr__(self) -> str:
...
+        >>> rng = torch.manual_seed(0)
+        >>> input = torch.rand(1, 1, 5, 5)
+        >>> sharpness = RandomSharpness(1.)
+        >>> sharpness(input)
+        tensor([[[[0.4963, 0.7682, 0.0885, 0.1320, 0.3074],
+                  [0.6341, 0.7720, 0.9537, 0.7566, 0.6323],
+                  [0.3489, 0.7325, 0.5629, 0.6284, 0.2939],
+                  [0.5185, 0.8648, 0.9106, 0.6249, 0.2823],
+                  [0.6816, 0.9152, 0.3971, 0.8742, 0.4194]],
+        <BLANKLINE>
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:901]: It is for passing the doctest. It will be rendered as ```<BLANKLINE>``` notation as here.

File: kornia/augmentation/functional.py [hunk 12]
```diff
@@ -518,10 +522,10 @@ def compute_affine_transformation(input: torch.Tensor, params: Dict[str, torch.T
 
 def apply_rotation(input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
     r"""Rotate a tensor image or a batch of tensor images a random amount of degrees.
-    Input should be a tensor of shape (C, H, W) or a batch of tensors :math:`(*, C, H, W)`.
+    Input should be a tensor of shape (C, H, W) or a batch of tensors :math:`(B, C, H, W)`.
 
     Args:
-        input (torch.Tensor): input image.
+        input (torch.Tensor): Tensor to be transformed with shape (H, W), (C, H, W), (B, C, H, W).
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/functional.py:528]: For either HW, CHW, BCHW, it will all be transformed into BCHW before applying any transformation.

File: kornia/augmentation/utils.py [hunk 13]
```diff
@@ -87,6 +77,10 @@ def _adapted_uniform(shape: Union[Tuple, torch.Size], low, high, same_on_batch=F
     If same_on_batch is True, all values generated will be exactly same given a batch_size (shape[0]).
     By default, same_on_batch is set to False.
     """
+    if not isinstance(low, torch.Tensor):
+        low = torch.tensor(low).float()
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/utils.py:81]: It is actually because the ```Uniform(low, high)``` function required the low and high to be float or double. An error will be raised if passing integer or long. So I think it is better to keep it this way for now.

File: kornia/color/adjust.py [hunk 14]
```diff
@@ -225,6 +227,248 @@ def adjust_brightness(input: torch.Tensor,
...
+        assert input.size(0) == len(thresholds) and len(thresholds.shape) == 1, \
+            f"threshholds must be a 1-d vector of shape ({input.size(0)},). Got {thresholds}"
+        # TODO: I am not happy about this line, but no easy to do batch-wise operation
+        thresholds = torch.stack([x.expand(*input.shape[1:]) for x in thresholds])
+
+    return torch.where(input < thresholds, input, 1.0 - input)
+
+
+def solarize(input: torch.Tensor, thresholds: Union[float, torch.Tensor] = 0.5,
+             additions: Optional[Union[float, torch.Tensor]] = 0.) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/color/adjust.py:258]: The hidden ```_solarize``` will not accept the ```additions``` parameter while I exposed the one with the ```additions``` parameter.

File: kornia/color/adjust.py [hunk 15]
```diff
@@ -225,6 +227,248 @@ def adjust_brightness(input: torch.Tensor,
     return out
 
 
+def _solarize(input: torch.Tensor, thresholds: Union[float, torch.Tensor] = 0.5) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/color/adjust.py:230]: The hidden ```_solarize``` will not accept the ```additions``` parameter while I exposed the one with the ```additions``` parameter.

File: kornia/color/adjust.py (hunk 2 above)
Comment (shijianjian) [reply kornia/color/adjust.py:283]: The outcome is the same. Using None could skip some calculations.

File: kornia/utils/image.py (hunk 8 above)
Comment (shijianjian) [reply kornia/utils/image.py:46]: Okay. I will make this one private for now. It is essentially doing nothing here but to convert either HW, CHW, HWC or BCHW into BCHW.
